pkg/registry/core/pod/strategy.go

```diff
@@ -569,3 +572,120 @@ func validateContainer(container string, pod *api.Pod) (string, error) {
+func applySeccompVersionSkew(pod *api.Pod) {
+	// initialize parents if necessary
+	if pod.Annotations == nil {
+		pod.Annotations = map[string]string{}
+	}
+
+	annotation, hasAnnotation := pod.Annotations[v1.SeccompPodAnnotationKey]
+	field, hasField := (*api.SeccompProfile)(nil), false
+	if pod.Spec.SecurityContext != nil && pod.Spec.SecurityContext.SeccompProfile != nil {
+		field = pod.Spec.SecurityContext.SeccompProfile
+		hasField = true
+	}
+
+	if hasAnnotation != hasField {
+		// assign computed values
+		newAnnotation, newField := syncSeccompAnnotationAndField(annotation, field)
+		if newAnnotation != "" {
+			pod.Annotations[v1.SeccompPodAnnotationKey] = newAnnotation
+		}
+		if newField != nil {
+			if pod.Spec.SecurityContext == nil {
+				pod.Spec.SecurityContext = &api.PodSecurityContext{}
+			}
+			pod.Spec.SecurityContext.SeccompProfile = newField
+		}
+	}
+
+	// Handle the containers of the pod
+	podutil.VisitContainers(&pod.Spec, podutil.AllFeatureEnabledContainers(),
+		func(ctr *api.Container, _ podutil.ContainerType) bool {
+			// find the annotation value
+			var (
+				annotation    string
+				hasAnnotation bool
+			)
+			for key, value := range pod.Annotations {
+				if key == api.SeccompContainerAnnotationKeyPrefix+ctr.Name {
+					annotation = value
+					hasAnnotation = true
+					break
+				}
+			}
+
+			// find the field
+			field, hasField := (*api.SeccompProfile)(nil), false
+			if ctr.SecurityContext != nil && ctr.SecurityContext.SeccompProfile != nil {
+				field = ctr.SecurityContext.SeccompProfile
+				hasField = true
+			}
+
+			if hasAnnotation != hasField {
+				// initialize parents if necessary
+				// assign computed values
+				newAnnotation, newField := syncSeccompAnnotationAndField(annotation, field)
+				if newAnnotation != "" {
+					pod.Annotations[v1.SeccompContainerAnnotationKeyPrefix+ctr.Name] = newAnnotation
+				}
+				if newField != nil {
+					if ctr.SecurityContext == nil {
+						ctr.SecurityContext = &api.SecurityContext{}
+					}
+					ctr.SecurityContext.SeccompProfile = newField
+
+				}
+			}
+
+			return true
+		})
+}
+
+// syncSeccompAnnotationAndField takes a pod annotation and a seccomp profile
+// field and returns the synchronized result: The annotation will be converted
+// in an appropriate profile and vice versa.
+func syncSeccompAnnotationAndField(annotation string, field *api.SeccompProfile) (string, *api.SeccompProfile) {
+	// If only seccomp fields are specified, add the corresponding annotations.
+	// This ensures that the fields are enforced even if the node version
+	// trails the API version
+	if field != nil && annotation == "" {
+		switch field.Type {
+		case api.SeccompProfileTypeUnconfined:
+			return v1.SeccompProfileNameUnconfined, nil
+
+		case api.SeccompProfileTypeRuntimeDefault:
+			return v1.SeccompProfileRuntimeDefault, nil
+
+		case api.SeccompProfileTypeLocalhost:
+			return v1.SeccompLocalhostProfileNamePrefix + *field.LocalhostProfile, nil
```

Comment:
this happens pre-validation, so field.LocalhostProfile can be nil... don't dereference without checking

Reply:
Added a check, in that case we now return `""` for the annotation (which is treated as invalid).